services/transaction.py

```diff
@@ -1,49 +1,191 @@
-from domain import block_details
+from typing import List, Tuple 
+from domain.gdk import GdkAPI, TransactionDetails
+from domain.locker import Locker
 from domain.receiver import Receiver
-from domain.utxo import CoinSelectionResult
+from domain.types import CoinSelectionResult, GdkUtxo, InputBlindingData, Outpoint, Utxo 
 import greenaddress as gdk
-from ocean.v1alpha.types_pb2 import BlockDetails
-from services.wallet import WalletService
+import wallycore as wally
+import secrets
+
+# For each output to blind, we need 32 bytes of entropy for each of:
+# - Output assetblinder
+# - Output amountblinder
+# - Ephemeral rangeproof ECDH key
+# - Explicit value rangeproof
+# - Surjectionproof seed
+OUTPUT_ENTROPY_SIZE = 5 * 32
 
 class TransactionService:
-    def __init__(self, wallet_svc: WalletService) -> None:
-        self._wallet_svc = wallet_svc
-    
+    def __init__(self, session: gdk.Session, locker: Locker) -> None:
+        self._session = session
+        self._gdk_api = GdkAPI(session)
+        self._locker = locker
+
     def sign_transaction(self, txHex: str) -> str:
         pass
-    
+
     def broadcast_transaction(self, txHex: str) -> str:
-        wallet = self._wallet_svc.get_wallet()
-        return gdk.broadcast_transaction(wallet.session, txHex)
-    
+        return self._gdk_api.broadcast(txHex)
+
+    def analyze_pset(self, psetBase64):
+        details = {
+            'psbt': psetBase64,
+            'utxos': self._gdk_api.get_all_unspents_outputs(),
+        }
+        return self._session.psbt_get_details(details)
+
+    def create_empty_pset(self) -> str:
+        pset = wally.psbt_init(2, 0, 0, 0, wally.WALLY_PSBT_INIT_PSET)
+        return wally.psbt_to_base64(pset, 0)
+
     def blind_pset(self, psetBase64: str) -> str:
-        wallet = self._wallet_svc.get_wallet()
-        pass
-    
+        psbt = wally.psbt_from_base64(psetBase64)
+        inputs_blinding_data = self._get_inputs_blinding_data(psetBase64)
+
+        valueBlindingFactors, values, assetBlindingFactors, assets = [wally.map_init(len(inputs_blinding_data), None) for _ in range(4)]
+        for i, blinding_data in enumerate(inputs_blinding_data):
+            wally.map_add_integer(values, i, blinding_data['value'])
+            wally.map_add_integer(valueBlindingFactors, i, blinding_data['value_blinder'])
+            wally.map_add_integer(assets, i, blinding_data['asset'])
+            wally.map_add_integer(assetBlindingFactors, i, blinding_data['asset_blinder'])
+
+        blinder_indexes_to_blind = set([u['input_index'] for u in inputs_blinding_data])
+        num_outputs = wally.psbt_get_num_outputs(psbt)
+        num_outputs_to_blind = 0
+        for i in range(num_outputs):
+            if wally.psbt_get_output_script_len(psbt, i) == 0:
+                continue # skip the fee outputs
+            out_blinder_index = wally.psbt_get_output_blinder_index(psbt, i)
+            if out_blinder_index in blinder_indexes_to_blind:
```

Comment:
here we should firstly check that the output has a blinding pubkey (= needs blinding) before this one